cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,957 @@
+  json_element_type cur_el_type;     // type of the current element
+  json_element_type parent_el_type;  // parent element type
+};
+
+enum class path_operator_type { ROOT, CHILD, CHILD_WILDCARD, CHILD_INDEX, ERROR, END };
+
+/**
+ * @brief A "command" operator used to query a json string.  A full query is
+ * an array of these operators applied to the incoming json string,
+ */
+struct path_operator {
+  CUDA_HOST_DEVICE_CALLABLE path_operator()
+    : type(path_operator_type::ERROR), index(-1), expected_type{NONE}
+  {
+  }
+  CUDA_HOST_DEVICE_CALLABLE path_operator(path_operator_type _type,
+                                          json_element_type _expected_type = NONE)
+    : type(_type), index(-1), expected_type{_expected_type}
+  {
+  }
+
+  path_operator_type type;  // operator type
+  // the expected element type we're applying this operation to.
+  // for example:
+  //    - you cannot retrieve a subscripted field (eg [5]) from an object.
+  //    - you cannot retrieve a field by name (eg  .book) from an array.
+  //    - you -can- use .* for both arrays and objects
+  // a value of NONE imples any type accepted
+  json_element_type expected_type;  // the expected type of the element we're working with
+  string_view name;                 // name to match against (if applicable)
+  int index;                        // index for subscript operator
+};
+
+/**
+ * @brief Parsing class that holds the current state of the JSONPath string to be parsed
+ * and provides functions for navigating through it. This is only called on the host
+ * during the preprocess step which builds a command buffer that the gpu uses.
+ */
+class path_state : private parser {
+ public:
+  path_state(const char* _path, size_t _path_len) : parser(_path, _path_len) {}
+
+  // get the next operator in the JSONPath string
+  path_operator get_next_operator()
+  {
+    if (eof()) { return {path_operator_type::END}; }
+
+    switch (*pos++) {
+      case '$': return {path_operator_type::ROOT};
+
+      case '.': {
+        path_operator op;
+        string_view term{".[", 2};
+        if (parse_path_name(op.name, term)) {
+          // this is another potential use case for __SPARK_BEHAVIORS / configurability
+          // Spark currently only handles the wildcard operator inside [*], it does
+          // not handle .*
+          if (op.name.size_bytes() == 1 && op.name.data()[0] == '*') {
+            op.type          = path_operator_type::CHILD_WILDCARD;
+            op.expected_type = NONE;
+          } else {
+            op.type          = path_operator_type::CHILD;
+            op.expected_type = OBJECT;
+          }
+          return op;
+        }
+      } break;
+
+      // 3 ways this can be used
+      // indices:   [0]
+      // name:      ['book']
+      // wildcard:  [*]
+      case '[': {
+        path_operator op;
+        string_view term{"]", 1};
+        bool const is_string = *pos == '\'' ? true : false;
```

Comment:
Only single-quotes are allowed in here?

Reply:
Yes...mostly.  The spec itself only allows for single-quotes, but there are some implementations (eg jsonpath_ng) which allow double quotes.  Kind of weird.